stats/memory_reporter.go

```diff
@@ -5,17 +5,26 @@ import (
 	"runtime"
 	"strconv"
 	"time"
+
+	"github.com/grafana/metrictank/util"
 )
 
 // MemoryReporter sources memory stats from the runtime and reports them
 // It also reports gcPercent based on the GOGC environment variable
 type MemoryReporter struct {
-	mem           runtime.MemStats
-	gcCyclesTotal uint32
+	mem                  runtime.MemStats
+	gcCyclesTotal        uint32
+	timeBoundGetMemStats func() interface{}
 }
 
 func NewMemoryReporter() *MemoryReporter {
-	return registry.getOrAdd("memory", &MemoryReporter{}).(*MemoryReporter)
+	reporter := registry.getOrAdd("memory", &MemoryReporter{}).(*MemoryReporter)
+	reporter.timeBoundGetMemStats = util.TimeBoundWithCacheFunc(func() interface{} {
+		mem := runtime.MemStats{}
+		runtime.ReadMemStats(&mem)
+		return mem
+	}, 5*time.Second, 1*time.Minute)
```

Comment:
After more discussion we think it will be best to model the timeout after 65% of the set interval. However, we don't know if this is enough time for the rest of the reporters to complete their operations. What are your thoughts on the best way to proceed with fixing this issue? Also of note, once we update it here we will need to update it in a few other projects.

Another option is to launch them all at the same time and wait for results. Any reporter that doesn't return a result within the interval will not get reported for that tick, but all the others will.
We can also add a budget to the launched reporting function and it can decide if it would like to use caching or not. We feel this is the better option, but it does add a bit of overhead which takes away from the overall time allocated for that tick. It still seems best. Thoughts?